synthtool/languages/java.py

```diff
@@ -551,3 +552,118 @@ class Example {
+    the end of the block by a closing brace at the same indentation
+    level. This requires the file to be correctly formatted.
+
+    Example: consider the following class:
+
+        class Example {
+            public void main(String[] args) {
+                System.out.println("Hello World");
+            }
+
+            public String foo() {
+                return "bar";
+            }
+        }
+
+    To copy and rename the `main` method above, use:
+
+        copy_and_rename_method('path/to/file', 'public void main(String[] args)', 'main', 'foo1')
+
+    Args:
+        filename (str): Path to source file
+        signature (str): Full signature of the method to remove. Example:
+            `public void main(String[] args)`.
+        before (str): name of the method to be copied
+        after (str): new name of the copied method
+    """
+    lines = []
+    method = []
+    leading_regex = None
+    with open(filename, "r") as fp:
+        line = fp.readline()
+        while line:
+            # for each line, try to find the matching
+            regex = re.compile("(\\s*)" + re.escape(signature) + ".*")
+            match = regex.match(line)
+            if match:
+                leading_regex = re.compile(match.group(1) + "}")
+                lines.append(line)
+                method.append(line.replace(before, after))
+                line = fp.readline()
+                continue
+
+            lines.append(line)
+            # not in a ignore block - preserve the line
+            if leading_regex:
+                method.append(line)
+            else:
+                line = fp.readline()
+                continue
+
+            # detect the closing tag based on the leading spaces
+            match = leading_regex.match(line)
+            if match:
+                # block is closed, resume capturing content
+                leading_regex = None
+                lines.append("\n")
+                lines.extend(method)
+
+            line = fp.readline()
+
+    with open(filename, "w") as fp:
+        for line in lines:
+            # print(line)
+            fp.write(line)
+
+
+def deprecate_method(filename: str, signature: str, alternative: str):
+    """Helper to deprecate a method.
+
+        Goes line-by-line to detect the start of the block.
+        Then adds the deprecation comment before the method signature.
+
+        Example: consider the following class:
+
+            class Example {
+                public void main(String[] args) {
+                    System.out.println("Hello World");
+                }
+
+                public String foo() {
+                    return "bar";
+                }
+            }
+
+        To deprecate the `main` method above, use:
+
+        DEPRECATION_WARNING = "This method is deprecated. Use {@link #{new_method}()} instead.
+        deprecate_method('path/to/file', 'public void main(String[] args)', DEPRECATION_WARNING.format(new_method="foo"))
```

Comment:
This doesn't match what the test is doing. The test has the user specify the leading open/close comment tags.

It would be nice if this detected any preceding javadoc and adds to the existing javadoc block, but if it's not possible, we should document the behavior with a caveat.

Reply:
Done